chronos/asyncloop.nim

```diff
@@ -796,25 +817,25 @@ proc withTimeout*[T](fut: Future[T], timeout: Duration): Future[bool] =
   ## future will hold false.
   var retFuture = newFuture[bool]("chronos.`withTimeout`")
   var moment: Moment
-  var timerPresent = false
+  var timer: TimerCallback
 
   proc continuation(udata: pointer) {.gcsafe.} =
     if not(retFuture.finished()):
       if isNil(udata):
         # Timer exceeded first.
-        fut.removeCallback(continuation)
+        clearTimer(timer)
```

Comment:
Why `fut.removeCallback` could be replaced with `clearTimer`?

Reply:
Because we need to maintain `addTimer()/removeTimer()` backwards compatible, I can probably add an overload `removeTimer` to take a `TimeoutCallback`, but it's cleaner to introduce a new pair of methods that makes things more evident.

Reply:
When timer completed first we are removing callback from Future[T], so it will be not called one more time. Your mistake here is that your clearing timer which is already completed.